lib/atx-heading.js

```diff
@@ -0,0 +1,315 @@
+    }
+
+    self.reconsume(CONTENT_STATE)
+  }
+}
+
+function contentState(code) {
+  var self = this
+  var info = self.contextInfo
+  var content = info.content
+
+  if (code === c.eof || code === c.lineFeed || code === c.carriageReturn) {
+    if (content) {
+      content.end = self.now()
+    }
+
+    self.reconsume(END_STATE)
+  } else if (code === c.space) {
+    if (content) {
+      content.end = self.now()
+    }
+
+    self.reconsume(CLOSING_SEQUENCE_BEFORE_STATE)
+  } else if (code === c.numberSign) {
+    if (content) {
+      content.end = self.now()
+    }
+
+    self.reconsume(CLOSING_SEQUENCE_STATE)
+  } else {
+    if (content === null) {
+      content = {start: self.now()}
+      info.content = content
+    }
+
+    self.consume()
+  }
+}
+
+function closingSequenceBeforeState(code) {
+  var self = this
+  var info = self.contextInfo
+  var before = info.closingSequenceBefore
+
+  if (code === c.eof || code === c.lineFeed || code === c.carriageReturn) {
+    if (before) {
+      before.end = self.now()
+    }
+
+    self.reconsume(END_STATE)
+  } else if (code === c.tab || code === c.space) {
+    if (info.closingSequenceBefore === null) {
+      before = {start: self.now()}
+      info.closingSequenceBefore = before
+    }
+
+    self.consume()
+  } else if (code === c.numberSign) {
+    if (before) {
+      before.end = self.now()
+    }
+
+    self.reconsume(CLOSING_SEQUENCE_STATE)
+  } else {
+    info.closingSequenceBefore = null
+    self.reconsume(CONTENT_STATE)
+  }
+}
+
+function closingSequenceState(code) {
+  var self = this
+  var info = self.contextInfo
+  var sequence = info.closingSequence
+
+  if (code === c.eof || code === c.lineFeed || code === c.carriageReturn) {
+    if (sequence) {
+      sequence.end = self.now()
+    }
+
+    self.reconsume(END_STATE)
+  } else if (code === c.tab || code === c.space) {
+    if (sequence) {
+      sequence.end = self.now()
+    }
+
+    self.reconsume(CLOSING_SEQUENCE_AFTER_STATE)
+  } else if (code === c.numberSign) {
+    if (sequence === null) {
+      sequence = {start: self.now()}
+      info.closingSequence = sequence
+    }
+
+    self.consume()
+  } else {
+    info.closingSequenceBefore = null
+    info.closingSequence = null
+    self.reconsume(CONTENT_STATE)
+  }
+}
+
+function closingSequenceAfterState(code) {
+  var self = this
+  var info = self.contextInfo
+  var after = info.closingSequenceAfter
+
+  if (code === c.eof || code === c.lineFeed || code === c.carriageReturn) {
+    if (after) {
+      after.end = self.now()
+    }
+    self.reconsume(END_STATE)
+  } else if (code === c.tab || code === c.space) {
+    if (after === null) {
+      after = {start: self.now()}
+      info.closingSequenceAfter = after
+    }
+
+    self.consume()
+  } else {
+    info.closingSequenceBefore = null
+    info.closingSequence = null
+    info.closingSequenceAfter = null
+    self.reconsume(CONTENT_STATE)
+  }
+}
+
+function bogusState() {
```

Comment:
Is this the equivalent thing for backtracking?

Reply:
Yes!